src/coprocessor/dag/expr/mod.rs

```diff
@@ -173,32 +178,75 @@ impl Expression {
                 ScalarFuncSig::NullEQJson => f.compare_json(ctx, row, f.sig),
                 _ => Err(Error::Other("Unknown signature")),
             },
-            _ => unimplemented!(),
         }
     }
 
     fn eval_real(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<f64>> {
-        unimplemented!()
+        match *self {
+            Expression::Constant(ref constant) => constant.eval_real(),
+            Expression::ColumnRef(ref column) => column.eval_real(row),
+            _ => unimplemented!(),
+        }
     }
 
-    fn eval_decimal(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<Decimal>> {
-        unimplemented!()
+    fn eval_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        match *self {
+            Expression::Constant(ref constant) => constant.eval_decimal(),
+            Expression::ColumnRef(ref column) => column.eval_decimal(row),
+            _ => unimplemented!(),
+        }
     }
 
-    fn eval_string(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<Vec<u8>>> {
-        unimplemented!()
+    fn eval_string<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Vec<u8>>>> {
+        match *self {
+            Expression::Constant(ref constant) => constant.eval_string(),
+            Expression::ColumnRef(ref column) => column.eval_string(row),
+            _ => unimplemented!(),
+        }
     }
 
-    fn eval_time(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<Time>> {
-        unimplemented!()
+    fn eval_time<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Time>>> {
+        match *self {
+            Expression::Constant(ref constant) => constant.eval_time(),
+            Expression::ColumnRef(ref column) => column.eval_time(row),
+            _ => unimplemented!(),
+        }
     }
 
-    fn eval_duration(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<Duration>> {
-        unimplemented!()
+    fn eval_duration<'a, 'b: 'a>(
```

Comment:
Here I believe it's necessary because in this arm
```rust
Expression::Constant(ref constant) => constant.eval_decimal(),
```
when we eval constant the lifetime is not relative with `row`.

And after remove `'b`, we will get compiling error.